zoom-us: fix memory usage bug, add test, update

Yarny0 opened the pull request:
*   Fix https://github.com/NixOS/nixpkgs/issues/371488 .
*   Update zoom (supersede https://github.com/NixOS/nixpkgs/pull/369514 ).
*   Add simple test for `zoom-us`.
*   Fix an error message about missing `pipewire`.

This is done with four commits (see their messages for more information):

*   Add a simple test derivation that starts zoom and checks for the zoom window to appear.  This should help catching bad updates.
*   Change the build recipe such that it is compatible with newer zoom versions: `zoom` eats all RAM as soon as it is touched by `patchelf`.  I haven't been able to uncover the reason for this.  However, by poking zoom inside a flatpak container and inside a `buildFHSEnv` container, it turned out it works well if called indirectly by calling the elf interpreter `ld-linux-x86-64.so.2` with the `zoom` binary as argument.  However, this generates a half-working zoom only -- e.g. it fails to play its usual notification sounds.  `strace` shows that it is looking for its files in the directory of the elf interpreter, because it assumes its own data files are stored where `/proc/self/exe` points to.  So we need to copy the elf interpreter in zooms data directory to make zoom find its files again.  Note that a symlink doesn't work -- apparently zoom follows that one.  Also settings argv0 doesn't help as this does not alter `/proc/self/exe`.
*   Update to newest zoom version: This simply cherry-picks the commit from https://github.com/NixOS/nixpkgs/pull/369514 .
*   `pipewire` is added to `PATH` to silence an error message.

Manipulating all those dynamic linking mechanisms felt awkward.  I guess this pull requests need thorough testing.

<details>
  <summary>If you want to test it inside a NixOS installation, you can use this overlay</summary>

```nix
config.nixpkgs.overlays = [( finalPkgs: prevPkgs: {
  zoom-us = prevPkgs.zoom-us.overrideAttrs (finalAttrs: prevAttrs: {
    version = "6.3.6.6315";
    src = finalPkgs.fetchurl {
      url = "https://zoom.us/client/${finalAttrs.version}/zoom_x86_64.pkg.tar.xz";
      hash = "sha256-QJR8SsMtyYBvd5G+mEjEEISkJJukCYeHErKrgs1uDQc=";
    };
    postFixup = finalPkgs.lib.replaceStrings
      [
        ''zopen zoom ZoomLauncher''
        ''makeWrapper $out/opt/zoom/.zoom $out/opt/zoom/zoom''
        '' PATH : ''
      ]
      [
        ''zopen ZoomLauncher''
        ''cp "$(cat $NIX_CC/nix-support/dynamic-linker)" $out/opt/zoom/ld.so ; makeWrapper $out/opt/zoom/{ld.so,zoom} --add-flags $out/opt/zoom/.zoom''
        '' PATH : ${finalPkgs.lib.makeBinPath [ finalPkgs.pipewire ]}:''
      ]
      prevAttrs.postFixup
    ;
  });
})];
```

</details>

Notifying maintainers: @danbst @tadfisher


## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [25.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) (or backporting [24.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) and [25.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

More (tested with the overlay stated above, on NixOS 24.11, X11, Plasma5, PulseAudio):

- [x] sending camera picture in video conference works
- [x] receiving camera picture in video conference works
- [x] audio is "detected" by zoom: microphone VU-meter shows amplitude
- [x] audio is produced: Test speaker generates audible ting-a-ling
- [x] audio is received in video conference
- [x] video is received in video conference
- [x] screen sharing sends screen to other participants
- [x] screen sharing of other participants is shown (tested without `pkgs.pipewire` in `PATH`)
- [x] reading team chat messages works as expected
- [x] SSO works: Zoom starts Firefox, then Firefox calls back to Zoom after login, as usual

`nixpkgs-review` tries to rebuild 6269 packages and was therefore skipped.

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

r-ryantm committed:
zoom-us: 6.2.11.5069 -> 6.3.6.6315

Yarny0 committed:
zoom-us (x86-64): don't touch the core `zoom` binary

Without the commit at hand, zoom-us stops working
and instead simply maxes out RAM when updated
from version 6.2.11.5069 to version 6.3.0.5527 .

https://github.com/NixOS/nixpkgs/issues/371488

Some experimenting yielded these observations
regarding newer (problematic) zoom-us versions on NixOS:
* They still work if started via flatpak
  (which creates a virtual FHS-compatible environment)
* They still work if started with `buildFHSEnv`
  with all libraries provided in `multiPkgs`.
* They still work if the unmodified zoom binary is started
  by manually invoking the `ld-linux` elf interpreter.
  However, this again requires that libraries
  are available via `LD_PRELOAD_PATH`.

It seems very much that the `patchelf --set-interpreter`
call on the `zoom` binary breaks something inside the binary.
The commit at hand removes `zoom` from the
list of binaries that are to be patched.
To ensure that program can still be executed, the `makeWrapper`
call (which is already there to wrap `zoom`) is extended
such that is explicitely uses the elf interpreter.

Unfortunatelly, when zoom tries to find its data
files (e.g. ring tones), it determines its
data directory by looking at `/proc/self/exe`.
If we start zoom by calling the elf interpreter,
`/proc/self/exe` acutally points to that interpreter.
The only solution is to copy the
interpreter into the zoom data directory
(a symlink isn't enough here).

Yarny0 committed:
zoom-us: add simple test

This test starts zoom with an Xvfb X server and
verifies that the "Zoom Workplace" window is created.
zoom is started with `prlimit` to avoid problems like
https://github.com/NixOS/nixpkgs/issues/371488 .

Note: During my experiments, it turned out that if certain
libraries are missing (I don't recall which ones),
then zoom still opens its Workplace window,
then immediatelly closes it again.
To reduce the likelihood of the test passing in
that scenario, another "sleep 20" is included.

philiptaron commented:
Yay, it doesn't consume all the memory!

Running the test I see this:

```
QSG_RENDER_LOOP is
XDG_CURRENT_DESKTOP = ;  DESKTOP_SESSION = ; XDG_SESSION_TYPE =
ls: cannot access '/usr/share/xdg-desktop-portal/portals/': No such file or directory
sh: pipewire: not found
[xdg-desktop-portal info]:
   can not find 'xdg-desktop-portal' command
   can not find 'xdg-desktop-portal-gnome' command
   can not find 'xdg-desktop-portal-kde' command
   can not find 'xdg-desktop-portal-gtk' command
   can not find 'xdg-desktop-portal-hyprland' command
   can not find 'xdg-desktop-portal-lxqt' command
   can not find 'xdg-desktop-portal-wlr' command
   can not find 'xdg-desktop-portal-xapp' command
[portal info] :
[pipewire info] :
```

Can this be fixed with some more wrappers?

Yarny0 committed:
zoom-us: provide pipewire program to fix error message

Without `pipewire` being in PATH, zoom complains

> sh: pipewire: not found

Yarny0 commented on pkgs/by-name/zo/zoom-us/test.nix:
```diff
@@ -0,0 +1,41 @@
+      # if libraries are missing, the window still appears,
+      # but disappears again immediatelly; check for that too:
+      sleep 20
+      grep -q "Zoom Workplace" window-names
```
Damn!  You're right, @jerith666 .  I hope I fixed it now.

jerith666 replied:
Yep, looks good now -- thanks!

Yarny0 commented:
> Running the test I see this:
> 
> ```
> QSG_RENDER_LOOP is
> XDG_CURRENT_DESKTOP = ;  DESKTOP_SESSION = ; XDG_SESSION_TYPE =
> ls: cannot access '/usr/share/xdg-desktop-portal/portals/': No such file or directory
> sh: pipewire: not found
> [xdg-desktop-portal info]:
>    can not find 'xdg-desktop-portal' command
>    can not find 'xdg-desktop-portal-gnome' command
>    can not find 'xdg-desktop-portal-kde' command
>    can not find 'xdg-desktop-portal-gtk' command
>    can not find 'xdg-desktop-portal-hyprland' command
>    can not find 'xdg-desktop-portal-lxqt' command
>    can not find 'xdg-desktop-portal-wlr' command
>    can not find 'xdg-desktop-portal-xapp' command
> [portal info] :
> [pipewire info] :
> ```
> 
> Can this be fixed with some more wrappers?

Mostly no, unfortunatelly.  I added `${pkgs.xdg-desktop-portal}/libexec` (where `xdg-desktop-portal` resides) to the `PATH`, but zoom won't find it.  `strace` tells me zoom is looking via hard-coded paths like `/usr/..` for those files.

But I added the `pipewire` program to `PATH` and the `pipewire: not found` error is now replaced by a small pipewire status message.  I'm not using pipewire, so I can't tell if this actually improves pipewire compatibility.

Yarny0 commented:
(Note: I updated the overlay script to include `pipewire` in `PATH`, and I tested most of the things mentioned above in the PR text again)

philiptaron commented:
If no other committer gets to this, I'll test and merge tomorrow, as I don't have access to my main NixOS machine today.

The changes look OK to me and I am quite grateful to have this blocker behind us.

No doubt there is still some work to do in order to make an environment for Zoom such that it works equally well on NixOS as it does on (say) Fedora or Ubuntu.

philiptaron commented on pkgs/by-name/zo/zoom-us/package.nix:
```diff
@@ -220,6 +228,7 @@ stdenv.mkDerivation {
   dontPatchELF = true;
 
   passthru.updateScript = ./update.sh;
+  passthru.tests.startwindow = callPackage ./test.nix { };
```
Future PR: this should do something like this, so that it tests the derivation that it's attached to in the presence of overrides.

```suggestion
  passthru.tests.startwindow = callPackage ./test.nix {
    zoom-us = finalAttrs.finalPackage;
  };
```

philiptaron commented on pkgs/by-name/zo/zoom-us/test.nix:
```diff
@@ -0,0 +1,44 @@
+        fi
+        sleep 1
+      done
+      # if libraries are missing, the window still appears,
+      # but disappears again immediatelly; check for that too:
```
```suggestion
      # If libraries are missing, the window still appears,
      # but then disappears again immediately; check for that also.
```

philiptaron commented on pkgs/by-name/zo/zoom-us/test.nix:
```diff
@@ -0,0 +1,44 @@
+      function is_zoom_window_present {
+        echo
+        xwininfo -root -tree  \
+            | sed 's/.*0x[0-9a-f]* \"\([^\"]*\)\".*/\1/; t; d'  \
```
golly

philiptaron reviewed (approved):
I tested this out, and the addition of pipewire enabled me to screenshare! There's more work to do, I think, because I did get a hang when leaving a meeting, and I was only able to screenshare _once_: the second time all that was shared was a black screen. Super finicky and buggy -- thanks Zoom. We'll work it out, and getting able to _run_ at all is a big win. I'll merge this PR. Thanks for working on it, @yarny0, and ping me on future PRs if you keep at it.
